webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala

```diff
@@ -521,6 +416,183 @@ object ConstructResponseUtilV2 {
+                                    case (pred: SmartIri, objs: Seq[LiteralV2]) => objs.map {
+                                        obj => pred -> obj
+                                    }
+                                }
+                        }
+
+                        // Flatten the value's statements.
+                        val valueStatements: FlatPredicateObjects = valueRdfWithUserPermission.assertions.flatMap {
+                            case (pred: SmartIri, objs: Seq[LiteralV2]) => objs.map {
+                                obj => pred -> obj
+                            }
+                        }
+
+                        // Get the rdf:type of the value.
+                        val rdfTypeLiteral: LiteralV2 = valueStatements.getOrElse(OntologyConstants.Rdf.Type.toSmartIri, throw InconsistentTriplestoreDataException(s"Value $valObjIri has no rdf:type"))
+
+                        val valueObjectClass: SmartIri = rdfTypeLiteral.asIriLiteral(
+                            throw InconsistentTriplestoreDataException(s"Unexpected object of $valObjIri rdf:type: $rdfTypeLiteral")
+                        ).value.toSmartIri
+
+                        // check if it is a link value
+                        if (valueObjectClass.toString == OntologyConstants.KnoraBase.LinkValue) {
+                            // create a link value object
+                            Some(
+                                ValueRdfData(
+                                    subjectIri = valObjIri,
+                                    valueObjectClass = valueObjectClass,
+                                    userPermission = valueRdfWithUserPermission.maybeUserPermission.get,
+                                    assertions = valueStatements,
+                                    standoff = emptyFlatStatements // link value does not contain standoff
+                                )
+                            )
+
+                        } else {
+                            // create a non-link value object
+                            Some(
+                                ValueRdfData(
+                                    subjectIri = valObjIri,
+                                    valueObjectClass = valueObjectClass,
+                                    userPermission = valueRdfWithUserPermission.maybeUserPermission.get,
+                                    assertions = valueStatements,
+                                    standoff = standoffAssertions
+                                )
+                            )
+                        }
+                }
+
+                property -> valueRdfData
+        }.filterNot {
+            // filter out those properties that do not have value objects (they may have been filtered out because the user does not have sufficient permissions to see them)
+            case (_, valObjs: Seq[ValueRdfData]) =>
+                valObjs.isEmpty
+        }
+
+        return valuePropertyToValueObject
```

Comment:
no `return` statement needed in Scala, see comment above.

Reply:
> Please find the details of my review in my comments.
> 
> In case you have questions regarding semantics (i.e. to describe what the purpose of some parts of the code is), we can talk next week. I barely remember what we did. So I figure it must be difficult for someone else. ;-)
> 
> But the method `splitMainResourcesAndValueRdfData` looks less scary now which is great! :-)

Would be great. I was not always quite sure what some parts were doing so I thought I would be better to keep the descriptions short to be sure it is not wrong.
Also with the return statement I am so used to always use it I completely forgot that in scala it is not needed :-)
And of course I will do the changes you marked.

Reply:
@nemmK 

> I was not always quite sure what some parts were doing

Then please get @tobiasschweizer to explain it, because we really need to make it clear in the code/comments.